python/pyarrow/compute.py

```diff
@@ -197,16 +215,13 @@ def _wrap_function(name, func):
     args_sig = ', '.join(arg_names)
     kwonly = '' if arg_names[-1].startswith('*') else ', *'
 
-    # Generate templated wrapper, so that the signature matches
-    # the documented argument names.
-    ns = {}
-    if option_class is not None:
-        template = _wrapper_options_template
-    else:
-        template = _wrapper_template
-    exec(template.format(func_name=name, args_sig=args_sig, kwonly=kwonly),
-         globals(), ns)
-    wrapper = ns['make_wrapper'](func, option_class)
+    # We make a generic wrapper with a simple argument-handling
+    # machinery, but we also add a __wrapped__ attribute pointing
+    # to a dummy callable with the desired signature for introspection
+    # and documentation.
+    wrapper = _make_generic_wrapper(name, func, option_class)
+    wrapper.__wrapped__ = _make_dummy_equivalent(args_sig, kwonly,
+                                                 option_class)
```

Comment:
While I do see how it can work, I'm not sure I got the benefit of building a dummy function with the right signature wrapped in the real function with the wrong signature.

I mean, if you are building the function with the right signature in `make_dummy_equivalent`, while not make that one not dummy, have it invoke the real compute function and directly return it?

The only reason I could think of is that in reality the signature you want to expose is not the real one you want to use.

Reply:
> The only reason I could think of is that in reality the signature you want to expose is not the real one you want to use.

This is exactly the case here: the function definition mostly forwards `*args` and `**kwargs` to the function kernel and the function options constructor, respectively. See the `wrapper` definitions above.

Reply:
The other benefit here is to confine the scope of `eval()`: now instead of using it to create the whole wrapper function we're only using it to get the signature right

Reply:
Uhm, if our only goal is to fake the signature, why not just use ``__signature__`` or ``__text_signature__`` which are available to expose signatures of C functions?

Like

```
def create_function(f_name, args):
    real_func = getattr(__builtins__, f_name)
    def f_impl(*args, **kwargs):
        return real_func(*args)
    f_impl.__name__ = f_name
    f_impl.__text_signature__ = "(a, b)"
    return f_impl

import inspect
f = create_function("min", ("a", "b"))
print(f.__name__, inspect.signature(f), f(1, 2))
```

That seem to achieve the same goal without any compile or ``__wrapped__`` magic

Reply:
`__text_signature__` is not really specified AFAICT, it's mostly for internal use by Argument Clinic (though other C generators may use it). We can try to set `__signature__`, however.

Reply:
Ok, it worked with `__signature__`. Thanks for the suggestion!